doc/tutorials/gpu-passthru.rst

```diff
@@ -0,0 +1,147 @@
+
+Pass-through GPU to Guest
+**************************
+
+1. Copy ``/usr/share/acrn/samples/nuc/launch_win.sh`` to ``install_win.sh``
+
+::
+
+   cp /usr/share/acrn/samples/nuc/launch_win.sh ~/install_win.sh
+
+2. Modify ``install_win.sh`` script to specify the Windows image you use.
+
+3. Modify ``install_win.sh`` script to enable GVT-d.
+
+Add the following commands before ``acrn-dm -A -m $mem_size -s 0:0,hostbridge \``
+
+::
+
+      gpudevice=`cat /sys/bus/pci/devices/0000:00:02.0/device`
+
+      echo "8086 $gpudevice" > /sys/bus/pci/drivers/pci-stub/new_id
+      echo "0000:00:02.0" > /sys/bus/pci/devices/0000:00:02.0/driver/unbind
+      echo "0000:00:02.0" > /sys/bus/pci/drivers/pci-stub/bind
+
+Replace ``-s 2,pci-gvt -G "$2" \`` with ``-s 2,passthru,0/2/0,gpu \``
+
+4. Run the ``launch_win.sh``.
+
+      .. note:: If you want to pass-through GPU to Clear Linux User VM, the steps are
+         the same as above except your script.
+
+Enable GVT-d GOP driver
+***********************
+
+Issue
+======
+When enabling GVT-d, guest OS couldn’t light up physical screen
+before OS driver load. So guest BIOS and
+grub GUI couldn’t be showed on the physical screen.
+
+Reason
+==========
+Physical display is initialized by GOP driver or VBIOS
+before OS driver load, but guest BIOS doesn’t have them.
+
+Solution
+==========
+Integrate GOP driver binary into OVMF as a Dxe driver,
+then guest OVMF could see GOP driver and run it in
+graphic pass-through environment.
+So physical display could be initialized
+by GOP and used by guest BIOS and guest grub.
+
+Steps
+======
+1. fetch ACRN OVMF
+
+::
+
+      git clone https://github.com/projectacrn/acrn-edk2.git
+
+2. fetch vbt and gop driver
+
+Fetch vbt and gop driver from the vendor according to your CPU model name.
+
+3. git apply the following two patches
+
+* `Use-the-default-vbt-released-with-GOP-driver.patch
+  <../_static/downloads/Use-the-default-vbt-released-with-GOP-driver.patch>`_
+* `Integrate-IntelGopDriver-into-OVMF.patch
+  <../_static/downloads/Integrate-IntelGopDriver-into-OVMF.patch>`_
+
+4. add vbt and gop driver into OVMF
+
+::
+
+      cp IntelGopDriver.efi  acrn-edk2/OvmfPkg/IntelGop/IntelGopDriver.efi
+      cp Vbt.bin acrn-edk2/OvmfPkg/Vbt/Vbt.bin
+
+5. compile OVMF
+
+::
+
+    cd edk2
+    git submodule update --init CryptoPkg/Library/OpensslLib/openssl
+
+    source edksetup.sh
+    make -C BaseTools
+
+    vim Conf/target.txt
+
+      ACTIVE_PLATFORM = OvmfPkg/OvmfPkgX64.dsc
+      TARGET_ARCH = X64
+      TOOL_CHAIN_TAG = GCC5
+
+    build -DFD_SIZE_2MB -DDEBUG_ON_SERIAL_PORT=TRUE
+
+Notes:
+
+   -  ``source edksetup.sh``, this step is needed for compilation every time
+      a shell is created.
+
+   -  This will generate the binary at
+      ``Build/OvmfX64/DEBUG_GCC5/FV/OVMF.fd``
+
+   -  Make sure your GCC is 5.X. GCC 6 and above is NOT supported.
```

Comment:
That means we need a dedicated machine to build this on and it cannot be the same as the one building ACRN. Is that correct? If so, let's make it more explicit.

Reply:
Yes, We need a dedicated machine to build. I will add more discription here.

Reply:
It can be the same one building ACRN. Just ensure GCC is 5.X.

Reply:
We don't want to recommend this, all our validation is done on systems that use much newer gcc versions. We even specifically ask users to manually upgrade gcc in an older version of Ubuntu.

@deb-intel @dbkinder , any thought on how we can best express this in the doc? (via a note for example?)

Reply:
We may provide a docker image that can successfully compile OVMF?

Reply:
It's a great idea if we can do that easily.

Reply:
We may refer [Build ACRN in Docker](https://projectacrn.github.io/latest/tutorials/building_acrn_in_docker.html).

[Compile OVMF with secure boot support](https://projectacrn.github.io/latest/tutorials/waag-secure-boot.html#compile-ovmf-with-secure-boot-support) also need to compile OVMF.

Reply:
> We may refer [Build ACRN in Docker](https://projectacrn.github.io/latest/tutorials/building_acrn_in_docker.html).
> 

That container uses a much newer gcc, have you tried it?

```$ sudo docker run acrn/clearlinux-acrn-builder bash -c "gcc -v"
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib64/gcc/x86_64-generic-linux/9/lto-wrapper
Target: x86_64-generic-linux
Configured with: ../gcc-9.3.0/configure --prefix=/usr --with-pkgversion='Clear Linux OS for Intel Architecture' --libdir=/usr/lib64 --enable-libstdcxx-pch --libexecdir=/usr/lib64 --with-system-zlib --enable-shared --enable-gnu-indirect-function --disable-vtable-verify --enable-threads=posix --enable-__cxa_atexit --enable-plugin --enable-ld=default --enable-clocale=gnu --disable-multiarch --enable-multilib --enable-lto --disable-werror --enable-linker-build-id --build=x86_64-generic-linux --target=x86_64-generic-linux --enable-languages=c,c++,fortran,go --enable-bootstrap --with-ppl=yes --with-isl --includedir=/usr/include --exec-prefix=/usr --with-glibc-version=2.19 --disable-libunwind-exceptions --with-gnu-ld --with-tune=haswell --with-arch=westmere --enable-cet --disable-libmpx --with-gcc-major-version-only --enable-default-pie
Thread model: posix
gcc version 9.3.1 20200416 releases/gcc-9.3.0-143-gbaf3a5a942 (Clear Linux OS for Intel Architecture) 

```

> [Compile OVMF with secure boot support](https://projectacrn.github.io/latest/tutorials/waag-secure-boot.html#compile-ovmf-with-secure-boot-support) also need to compile OVMF.

Good point... it'd be good to update that doc too if we can provide a simple mechanism.

Reply:
> > We may refer [Build ACRN in Docker](https://projectacrn.github.io/latest/tutorials/building_acrn_in_docker.html).
> 
> That container uses a much newer gcc, have you tried it?
> 
> ```
> Using built-in specs.
> COLLECT_GCC=gcc
> COLLECT_LTO_WRAPPER=/usr/lib64/gcc/x86_64-generic-linux/9/lto-wrapper
> Target: x86_64-generic-linux
> Configured with: ../gcc-9.3.0/configure --prefix=/usr --with-pkgversion='Clear Linux OS for Intel Architecture' --libdir=/usr/lib64 --enable-libstdcxx-pch --libexecdir=/usr/lib64 --with-system-zlib --enable-shared --enable-gnu-indirect-function --disable-vtable-verify --enable-threads=posix --enable-__cxa_atexit --enable-plugin --enable-ld=default --enable-clocale=gnu --disable-multiarch --enable-multilib --enable-lto --disable-werror --enable-linker-build-id --build=x86_64-generic-linux --target=x86_64-generic-linux --enable-languages=c,c++,fortran,go --enable-bootstrap --with-ppl=yes --with-isl --includedir=/usr/include --exec-prefix=/usr --with-glibc-version=2.19 --disable-libunwind-exceptions --with-gnu-ld --with-tune=haswell --with-arch=westmere --enable-cet --disable-libmpx --with-gcc-major-version-only --enable-default-pie
> Thread model: posix
> gcc version 9.3.1 20200416 releases/gcc-9.3.0-143-gbaf3a5a942 (Clear Linux OS for Intel Architecture) 
> ```
> 
> > [Compile OVMF with secure boot support](https://projectacrn.github.io/latest/tutorials/waag-secure-boot.html#compile-ovmf-with-secure-boot-support) also need to compile OVMF.
> 
> Good point... it'd be good to update that doc too if we can provide a simple mechanism.

I didn't try it.  You may ask nanlin and wenling have a try, thanks.

Reply:
> I didn't try it. You may ask nanlin and wenling have a try, thanks.

I gave this a (quick) try. But it did not work. I suggest we put a clear note explaining that the user will need a different build machine that has a gcc 5.x version installed.

Reply:
> > I didn't try it. You may ask nanlin and wenling have a try, thanks.
> 
> I gave this a (quick) try. But it did not work. I suggest we put a clear note explaining that the user will need a different build machine that has a gcc 5.x version installed.

Ok, I will put a clear note explaining that the user will need a different build machine that has a gcc 5.x version installed

Reply:
> > > I didn't try it. You may ask nanlin and wenling have a try, thanks.
> > 
> > 
> > I gave this a (quick) try. But it did not work. I suggest we put a clear note explaining that the user will need a different build machine that has a gcc 5.x version installed.
> 
> Ok, I will put a clear note explaining that the user will need a different build machine that has a gcc 5.x version installed

We don't need to emphasize `different`.  **You need a build machine that has a GCC 5.X version installed** is enough.

Reply:
OK, make it a clear note though.